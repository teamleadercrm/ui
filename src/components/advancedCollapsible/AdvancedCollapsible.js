import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { TextBody, Heading3 } from '../typography';
import Icon from '../icon';
import Box, { pickBoxProps } from '../box';
import { IconChevronDownSmallOutline, IconChevronRightSmallOutline } from '@teamleader/ui-icons';
import theme from './theme.css';

const AdvancedCollapsible = ({ children, color, size, title, ...others }) => {
  const [collapsed, setCollapsed] = useState(true);

  const boxProps = pickBoxProps(others);
  const TitleElement = size === 'large' ? Heading3 : TextBody;

  const handleTitleClick = () => {
    setCollapsed(!collapsed);
  };

  return (
    <Box>
      <Box className={theme['title']} display="flex" alignItems="center" onClick={handleTitleClick} {...boxProps}>
        <Icon color={color} tint="darkest">
          {collapsed ? <IconChevronRightSmallOutline /> : <IconChevronDownSmallOutline />}
        </Icon>
        <TitleElement color={color} marginLeft={2} tint="darkest">
          {size === 'medium' ? <strong>{title}</strong> : title}
        </TitleElement>
      </Box>
      {!collapsed && (
        <Box className={theme['children']} marginTop={2}>
          {children}
        </Box>
      )}
    </Box>
  );
};

AdvancedCollapsible.propTypes = {
  color: PropTypes.oneOf(['neutral', 'teal']),
  children: PropTypes.any,
  title: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

AdvancedCollapsible.defaultProps = {
  color: 'teal',
  size: 'medium',
};

export default AdvancedCollapsible;
