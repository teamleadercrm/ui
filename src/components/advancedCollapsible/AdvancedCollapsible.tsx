import React, { ReactNode, useState } from 'react';
import { TextBody, Heading3 } from '../typography';
import Icon from '../icon';
import Box, { pickBoxProps } from '../box';
import { IconChevronDownSmallOutline, IconChevronRightSmallOutline } from '@teamleader/ui-icons';
import theme from './theme.css';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';
import { COLORS, SIZES } from '../../constants';

export type AllowedAdvancedCollapsibleSize = Exclude<typeof SIZES[number], 'tiny' | 'fullscreen' | 'smallest' | 'hero'>;
export type AllowedAdvancedCollapsibleColor = Exclude<
  typeof COLORS[number],
  'aqua' | 'gold' | 'mint' | 'ruby' | 'violet'
>;
export interface AdvancedCollapsibleProps extends Omit<BoxProps, 'size'> {
  color?: AllowedAdvancedCollapsibleColor;
  children: ReactNode;
  title: string;
  size?: AllowedAdvancedCollapsibleSize;
}

const AdvancedCollapsible: GenericComponent<AdvancedCollapsibleProps> = ({
  children,
  color = 'teal',
  size = 'medium',
  title,
  ...others
}) => {
  const [collapsed, setCollapsed] = useState(true);

  const boxProps = pickBoxProps(others);
  const TitleElement = size === 'large' ? Heading3 : TextBody;

  const handleTitleClick = () => {
    setCollapsed(!collapsed);
  };

  return (
    <Box>
      <Box className={theme['title']} display="flex" alignItems="center" onClick={handleTitleClick} {...boxProps}>
        <Icon color={color} tint="darkest">
          {collapsed ? <IconChevronRightSmallOutline /> : <IconChevronDownSmallOutline />}
        </Icon>
        <TitleElement color={color} marginLeft={2} tint="darkest">
          {size === 'medium' ? <strong>{title}</strong> : title}
        </TitleElement>
      </Box>
      {!collapsed && (
        <Box className={theme['children']} marginTop={2}>
          {children}
        </Box>
      )}
    </Box>
  );
};

export default AdvancedCollapsible;
