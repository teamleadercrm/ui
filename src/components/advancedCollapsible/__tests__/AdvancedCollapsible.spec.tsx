import React from 'react';
import { render } from '@testing-library/react';
import AdvancedCollapsible from '../AdvancedCollapsible';

describe('Component - AdvancedCollapsible', () => {
  it('renders', async () => {
    const { asFragment } = render(<AdvancedCollapsible title="Foo">Bar</AdvancedCollapsible>);
    expect(asFragment()).toMatchSnapshot();
  });
});
