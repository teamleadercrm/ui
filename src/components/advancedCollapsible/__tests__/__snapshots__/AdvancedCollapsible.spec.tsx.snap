// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - AdvancedCollapsible renders 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <div
      class="box align-items-center display-flex title"
    >
      <span
        class="box align-items-center display-flex teal darkest"
        data-teamleader-ui="icon"
      >
        <svg
          fill="currentColor"
          height="14"
          preserveAspectRatio="xMidYMid meet"
          style="opacity: 0.84; vertical-align: middle;"
          viewBox="0 0 14 14"
          width="14"
        >
          <path
            clip-rule="evenodd"
            d="M6.086 10.537L9.62 7 6.086 3.466A1 1 0 104.672 4.88l2.12 2.121-2.12 2.122a1 1 0 101.414 1.414z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
      <p
        class="box margin-left-2 text text-body darkest teal"
        data-teamleader-ui="text"
      >
        <strong>
          Foo
        </strong>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
